packages/devextreme/js/__internal/ui/selection/m_selection.strategy.ts

```diff
@@ -33,6 +33,53 @@ export default class SelectionStrategy {
     this.options[name] = value;
   }
 
+  onSelectionChanging(): boolean | Promise<boolean> {
+    const {
+      selectedItems,
+      selectedItemKeys,
+      addedItemKeys,
+      removedItemKeys,
+      addedItems,
+      removedItems,
+      onSelectionChanging = noop,
+    } = this.options;
+
+    const selectionChangingArgs = {
+      selectedItems,
+      selectedItemKeys,
+      addedItemKeys,
+      removedItemKeys,
+      addedItems,
+      removedItems,
+      cancel: false,
+    };
+
+    onSelectionChanging(selectionChangingArgs);
+    return selectionChangingArgs.cancel;
+  }
+
+  _callCallbackIfNotCanceled(callback: () => void, cancelCallback: () => void): void {
+    const cancelResult = this.onSelectionChanging();
+
+    if (isPromise(cancelResult)) {
+      cancelResult
+        .then((cancel) => {
+          if (!cancel) {
+            callback();
+          } else {
+            cancelCallback();
+          }
+        })
+        .catch(() => {
+          callback();
+        });
+    } else if (!cancelResult) {
+      callback();
+    } else {
+      cancelCallback();
+    }
```

Comment:
It's possible but not so easy to rewrite - in your suggestion part of cases is not considered.
And also it would be strange to await this function call in parent function which works with deferred